### PR TITLE
Add copyright.txt and test script

### PR DESCRIPTION
**Pull request by idotobi**
- add copyright.txt to repo using the debian copyright format
 - add script to easily check if something needs to be updated
   in the copyright.txt

### CODE REVIEW

**Comment by idotobi** [comment]
@rolandlo : This would be ready for review now.

The output of the script is:

```sh
$ python scripts/compare_license.py 
Found License/Copyright both in copyright.txt and repo:  22
- All automatically detected files listed or whitelisted
- All listed files automatically detected or whitelisted
- No listed file got changed since the last check.
🎉 Success
```

@barak : Does this look good to you as well? Would you be alright with me squashing the commit when merging?

File: copyright.txt
Comment (idotobi) [inline copyright.txt:32]: @Fatih20 : Is this OK for you?

File: copyright.txt
Comment (rolandlo) [inline copyright.txt:35]:
```suggestion
         These icons where heavily inspired by https://github.com/elementary/icons
```

File: copyright.txt
Comment (rolandlo) [inline copyright.txt:21]: What is our scheme for the years here? Some only have a beginning, others have a period.

File: scripts/compare_license.py
Comment (rolandlo) [inline scripts/compare_license.py:1]: Maybe this script needs some licensing information for itself.

**Review by rolandlo — COMMENTED** [review]
This look pretty good. I will have to double check the licensing information given in the files. I am no licensing expert, so hopefully someone with more experience can take a look as well.

**Review by rolandlo — CHANGES_REQUESTED** [review]
I went through all the files listed in copyrights.txt and have requested some changes. Other than that it looks all good to me. Great that you wrote this python script! It will make things so much easier in the future.

File: copyright.txt
Comment (Febbe) [reply copyright.txt:139]: Hehe in 2003 I was in the elementary school so that date can't be true.

File: copyright.txt
Comment (Febbe) [reply copyright.txt:21]: Those with a beginning only are either not updated or only worked in that year.

File: copyright.txt
Comment (Febbe) [reply copyright.txt:82]: I was the one who created it, but it's a 1:1 copy of boost::hash_combine, so it's not my Copyright https://www.boost.org/doc/libs/1_72_0/doc/html/hash/reference.html#boost.hash_combine.

**Comment by Febbe** [comment]
Is the copyright.txt for the whole folder, or for the binary application?
If it's for the binary, we only need the copyright for those which are included in the release package.
So we wouldn't require the copyright for cmake files etc.

File: scripts/compare_license.py
Comment (idotobi) [reply scripts/compare_license.py:1]: :rofl: Nah it's alright. By default it gets licensed under the same license as the repo and I'm totally fine with this.

From what I read up, actually the copyright would even still remain with me without any information anywhere as long as I could proof that it comes from me (git for the win). But I'm not a lawyer and might be wrong in which case I'm still fine with it :nerd_face:

File: copyright.txt
Comment (idotobi) [reply copyright.txt:21]: The way I would understand it, is that:
* `yyyy, Name`: Person still part of active team
* `yyyy-yyyy, Name`: Person no longer part of active team

I think we have two options here:

1. Updating this list by taking the `About` info as source of truth (but being careful with removal of names)
2. Removing all entries and only keep `Xournal Team Members`.

What do you think @Febbe , @rolandlo ?

File: copyright.txt
Comment (idotobi) [reply copyright.txt:56]: You're completely right. Seems like I messed up the rebase :facepalm:  (see https://github.com/xournalpp/xournalpp/issues/3103#issuecomment-847029730). I'll fix that.

Comment (idotobi) [reply copyright.txt:56]: I just double checked this was the only thing I picked wrong :see_no_evil:

File: copyright.txt
Comment (idotobi) [reply copyright.txt:82]: @Febbe : Is the reference to Devid Eberly still valid then? The only thing I found through googling was this https://www.geometrictools.com/GTE/Mathematics/HashCombine.h.

File: copyright.txt
Comment (idotobi) [reply copyright.txt:126]: @rolandlo : To be honest I'm not sure. I noticed that in the xournalpp codebase no cpp file as a Licensing text (at least I didn't find any). So I assumed that this is somehow a convention in cpp programming to only put licensing/copyright on `.h` files. 
Do you have any experience with this?

File: copyright.txt
Comment (idotobi) [reply copyright.txt:139]: Way to start early :wink:  I checked the commit and updated it to 2020.

File: scripts/compare_license.py
Comment (idotobi) [reply scripts/compare_license.py:20]: :pray: Thanks!

**Comment by idotobi** [comment]
@Febbe  I don't really know but @barak might know what the scope of the `copyright.txt` file is in debian, which is where I took this format from.

In any way I think, once established, maintaining it should not take to much efforts. Especially once we add a contributing file explaining how copyright/licensing works by default in the repo, I would hope expect that only few files would have extra licensing.

File: copyright.txt
Comment (rolandlo) [reply copyright.txt:21]: > The way I would understand it, is that:
> 
>     * `yyyy, Name`: Person still part of active team
> 
>     * `yyyy-yyyy, Name`: Person no longer part of active team
> 

What about team members, that have only contributed for 1 year (in the past)? Would that be written like 2016-2016?
 
> I think we have two options here:
> 
>     1. Updating this list by taking the `About` info as source of truth (but being careful with removal of names)
> 
>     2. Removing all entries and only keep `Xournal Team Members`.

The term `Xournal Team Members` does not refer to the Xournal++ team, since Xournal is a different software. There are two files `ZoomCallib.h` and `LoadHandlerHelper.h` that have an `@author Xournal Team` line and refer to  http://xournal.sf.net. That's why `Xournal Team Members` is listed, I think.

> 
> 
> What do you think @Febbe , @rolandlo ?

The `About info` is outdated and it doesn't contain email addresses. A source of truth is https://github.com/xournalpp/xournalpp/graphs/contributors
If "work" means "commits to the master branch, the years/names/email addresses (ordered by the one who was a team member first to the newest team member) would appears to be:

           2012-2018, Wilson Brenna <wilson@wbrenna.homelinux.com>
           2013-2021, Andreas Butti <andreas.butti@gmail.com>
           2015-2016, Marek Pikuła <marek@pikula.tk>
           2017-2019, peetCreative <peetCreative@users.noreply.github.com>
           2018-2019, Moreno Razzoli <morrolinux@users.noreply.github.com>
           2018-2019, Luca Errani <luca.errani1@gmail.com>
           2019-2021, Ulrich Huber <ulrich@huberulrich.de>
           2019-2019, Justin Jones <bjustjones@netscape.net>
           2019-2021, Bryan Tan <Technius@users.noreply.github.com>
           2019-2019, Justus Rossmeier <veecue@users.noreply.github.com>
           2019-2021, Fabian Keßler <Fabian_Kessler@gmx.de>
           2019-2020, Julius Lehmann <internet@devpi.de>
           2020-2021, Tobias Hermann <idotobi@users.noreply.github.com>
           2020-2021, Roland Lötscher <roland_loetscher@hotmail.com>

I don't know the name of peetCreative, but he or she is a team member. There may have been other team members  (maybe @x2b 2013-2015) contributing to Xournal++, but I would not know of that for sure.

           2003, Ettore Perazzoli
           1995, Spencer Kimball and Peter Mattis

are not Xournal++ team members. They have contributed to GTK.

**Comment by barak** [comment]
The intent is for `debian/copyright` to cover all the material in the repo, period. The format is agnostic on that issue I suppose, but I don't see any disadvantage to covering everything. If someone wants a subset for some reason, the format would make it easy to automatically remove unnecessary stanzas.

File: copyright.txt
Comment (rolandlo) [reply copyright.txt:126]: @idotobi I have no experience with this. In comparison, [Inkscape](https://gitlab.com/inkscape/inkscape/-/tree/master/src) has a copyright notice in every `.cpp`-file and every `.h`-file.

File: copyright.txt
Comment (Febbe) [reply copyright.txt]: No he also copied it(but he created the file and referenced the cited paper). Maybe it is ok to give the Copyright of that file to me, since I created it and I just cited the code from boost, so we need to cite the boost source?

And the one who created the boost file is Daniel James (2005-2014)

File: scripts/compare_license.py
Comment (LittleHuba) [inline scripts/compare_license.py:113]: Could you modify this line to automatically get the commit hash of the last official release? This would simplify usage by quite a bit.

File: copyright.txt
Comment (LittleHuba) [reply copyright.txt:126]: There is definitely no such convention. Generally it is assumed that if there is no explicit copyright info in a file, the one of the repository root applies. So if it is not the same we must add the info to the file.

File: scripts/compare_license.py
Comment (idotobi) [reply scripts/compare_license.py:113]: In theory yes but, then the script would always exit with `1` if any file changed since last release, and there's no clear way to set the status to "rechecked, all good".

The workflow that is intended with the script is rather.

 - Run before next release
 - if it exits with 0: all good
 - if it exits with 1:
     - check the files that have been changed
     - update licenses if required
     - update `last_checked_git_commit_hash` to last git commit hash on master
     - merge change to master

This way when you release this script should always yield a `0` exit code, if not one shouldn't yet release.

Does that make sense, or would you have a different idea for the workflow?

File: copyright.txt
Comment (idotobi) [reply copyright.txt:126]: Thanks for the confirmation @rolandlo and @LittleHuba I'll go through the files to check if we miss any headers.

For this particular one I'd actually be tempted to remove the licensing/copyright header as there's almost no resemblence between the last commit by Wilson Brenna and the current state of the file (see commit 0dc7b6e5e0e31804b1e4d65dcfe512614529270e).

But I guess as this is not that important since the license in anyway is the same as the one of the main project I'll just add the copyright/license information also to the `.cpp` file, just to be on the safe side.

File: scripts/compare_license.py
Comment (LittleHuba) [reply scripts/compare_license.py:113]: Could we instead create a script that can be automatically run on a merge request and check if a license changed in a file or one was added without license? That would make it obvious to the person providing the merge request to add this info.

Comment (idotobi) [reply scripts/compare_license.py:136]: This is actually how the script could be used as well, however there's one caveat why it might make sense to only make it mandatory to run it pre release.

Following Usecase is problematic:
 1. You updated a file with license
 2. You run script on PR -> script raises warning
 3. You recheck and update the commit hash in the script to the last one of your PR
 4. You commit this change to your PR -> Now the PR builds without problems
 5. You squash merge 
 6. :boom: 

In this case the `last_checked_git_commit_hash` would no longer be valid and the script would fail.

I see two solutions for this problem:
 1. Do never squash PRs (probably not that great in the long run)
 2. Allow to merge PRs even if the script fails (it would then be expected/ideal if the respective developer would raise a new PR updating the `last_checked_git_commit_hash` based on his merged last PR)

For me Nr. 2 seems like the best way. However, I'm not sure if this is an easy enough workflow to expect that every contributor knows about this.

Hence, I would say for release builds the script should never fail (as the main contributors probably know how to fix it). 
For usual PRs it's just a warning but doesn't prevent a merge. In the worst case one would need to ask the person who merged without updating the `last_checked_git_commit_hash` for details prior to the release if something is unclear.

Comment (LittleHuba) [reply scripts/compare_license.py:136]: IFor PRs the commit hash should not be updated manually. This would equal a merge conflict for every PR we have.

Instead use HEAD of master as base commit and check for all files that are added/removed/changed whether the file has a license and the copyright data is accurate. We could enforce SPDX identifiers for this in all new/modified files.

File: copyright.txt
Comment (rolandlo) [inline copyright.txt:31]:
```suggestion
comment: The files were committed by Andreas Butti initially who was active for both
```

File: src/control/xojfile/LoadHandlerHelper.cpp
Comment (rolandlo) [inline src/control/xojfile/LoadHandlerHelper.cpp:4]:
```suggestion
 * helper methods to load a .xoj / .xopp document
```

File: copyright.txt
Comment (idotobi) [reply copyright.txt]: > What about team members, that have only contributed for 1 year (in the past)? Would that be written like 2016-2016?

Yes that's weird. We can of course also do it as you proposed. We just need to remember to update the end date (e.g. 2019-2021 will need to be updated in a year). But I guess that should be fine.
Also it's actually more consistent with the other dates.

> The term Xournal Team Members does not refer to the Xournal++ team, since Xournal is a different software.

Ah :facepalm: , I didn't notice the difference. I actually should have seen that in the script, but there was a subtle bug. Good that you pointed me to this otherwise I would have never seen that :pray: .

> I don't know the name of peetCreative, but he or she is a team member.

I added a comment to the `copyright.txt` file that for a complete and full contributor list one should refer to `git log`. I think/hope this should be fine.

**Review by rolandlo — CHANGES_REQUESTED** [review]
To me everything looks good now, except that the list/years of Xournal++ authors from the files `AUTHORS` and `about.glade` should be made to agree with the one from `copyright.txt`. 

I agree with @idotobi that running the script before a new release is all we really need.

**Comment by idotobi** [comment]
@rolandlo : Should we update the `AUTHORS` and `about.glade` in a different PR?

For the `about.glade` I would have two proposals:

1. To specifically mention the the two designers for the Icons.
2. To change the layout from 1-column to 2-column to prevent overflow for small screen sizes (aka. my laptop :sweat_smile: )
image

What do you think?

**Comment by rolandlo** [comment]
> @rolandlo : Should we update the `AUTHORS` and `about.glade` in a different PR?

Thinking of it, yes that may be better. Currently `AUTHORS` only lists the member of the maintainer team. Comparing with other projects it may be more appropriate to list all contributors, which is basically the output of 
```
git shortlog -s -n
```

> 
> For the `about.glade` I would have two proposals:
> 
>     1. To specifically mention the the two designers for the Icons.
> 
>     2. To change the layout from 1-column to 2-column to prevent overflow for small screen sizes (aka. my laptop sweat_smile )

Both 1. and 2. sound good to me.

**Review by rolandlo — APPROVED** [review]

**Comment by rolandlo** [comment]
For the AUTHORS file I found a nice little script: https://gist.github.com/vladikoff/d3f5ec1f570e66fb7faff329d01012b7

**Comment by rolandlo** [comment]
Some squashing is still needed, before this can be merged.

File: scripts/compare_license.py
Comment (LittleHuba) [reply scripts/compare_license.py:142]: For now this script is absolutely okay. Delaying the release it not worth this commit. My point is more for the long run.

Main goal for the future should be that no code gets submitted without correct copyright notice. Otherwise code from this specific commit is not covered by the copyright (from a very strict point of view). I know of projects that rebased the whole project because of this.

I'd recommend SPDX tags for concise and consistent copyright notices throughout all files. They are currently being established as kind of a standard throughout open source projects as more and more teams include them in their sources.

**Comment by rolandlo** [comment]
@idotobi Would you please squash your commits into reasonable pieces and rebase on current master? Otherwise, if no objections are raised, I will simply "Squash and merge" in 24 hours.

**Comment by idotobi** [comment]
@rolandlo : I was away for the weekend. Got back just in time :sweat_smile:  . I did the squash into two commits

1. Including all changes
2. Updating the last checked commit hash onto (1.)

Hence, the `compare_license.py` script now prints success and should also do so for release as long as the merge to master is now done without merging (to not loose the commit hash).

If you agree with this feel free to merge :nerd_face:

**Comment by rolandlo** [comment]
Wait a moment. For your PR I am supposed to "Create a merge commit", not "Squash and merge" (because there are 2 commits that should stay separated) and not "Rebase and merge" (because that would change commit hashes, see [here](https://stackoverflow.com/questions/60597400/how-to-do-a-fast-forward-merge-on-github)). But "Create a merge commit" is not enabled for this branch. So am I missing something or do we need to enable "Create a merge commit" somewhere?

**Comment by Febbe** [comment]
There is no reason, to keep the commit-hashes, if noone else based his work onto that branch. 
So doing a squash and rebase is completely fine here.

**Comment by rolandlo** [comment]
> There is no reason, to keep the commit-hashes, if no-one else based his work onto that branch so far.
> So doing a squash and rebase is completely fine here.
> Also, every modification of the history will modify the hashes. This means, that even the squash will change those.
> 
> I disabled the ability to merge, since it happened to often that this was used as default without thinking.

Are you sure? The second commit https://github.com/xournalpp/xournalpp/pull/3120/commits/5143cab4f15d576e7583848f2379b0c4b2151313 has the first commit hash hard-coded.

**Comment by Febbe** [comment]
> Are you sure? The second commit [5143cab](https://github.com/xournalpp/xournalpp/commit/5143cab4f15d576e7583848f2379b0c4b2151313) has the first commit hash hard-coded.

That is indeed another case, where you want to keep the history of the hard coded hash.
But now it is too late, the second commit must be rebased/amended again to change the commit hash.
It can also be helpful, to first merge the first commit into master, and then merge the second.

**Comment by Febbe** [comment]
@rolandlo creating a merge commit locally and merging it back should also work. To circumvent this.

**Comment by rolandlo** [comment]
@Febbe Is that something you could do? The only solution I fully understand is to "Squash and merge" this PR and then create a second PR which updates the `last_checked_git_commit_hash` in a new PR.

**Comment by Febbe** [comment]
dunno if I am allowed to force push to this branch

**Comment by rolandlo** [comment]
> dunno if I am allowed to force push to this branch

I don't know either, but I do not have more rights than you do.

**Comment by Febbe** [comment]
Thanks for doing this, hadn't time till now.

**Comment by idotobi** [comment]
@rolandlo : Thanks :pray:

**Pull request merged**
